<a href="https://colab.research.google.com/github/h8erqqq/ML/blob/main/PanelData.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [3]:
import numpy as np
import pandas as pd

In [4]:
df = pd.read_csv('/content/Financial Statements.csv')

df.info(), df.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 161 entries, 0 to 160
Data columns (total 23 columns):
 #   Column                               Non-Null Count  Dtype  
---  ------                               --------------  -----  
 0   Year                                 161 non-null    int64  
 1   Company                              161 non-null    object 
 2   Category                             161 non-null    object 
 3   Market Cap(in B USD)                 160 non-null    float64
 4   Revenue                              161 non-null    float64
 5   Gross Profit                         161 non-null    float64
 6   Net Income                           161 non-null    float64
 7   Earning Per Share                    161 non-null    float64
 8   EBITDA                               161 non-null    float64
 9   Share Holder Equity                  161 non-null    float64
 10  Cash Flow from Operating             161 non-null    float64
 11  Cash Flow from Investing        

(None,
    Year Company  Category  Market Cap(in B USD)   Revenue  Gross Profit  \
 0  2022     AAPL       IT               2066.94  394328.0      170782.0   
 1  2021     AAPL       IT               2913.28  365817.0      152836.0   
 2  2020     AAPL       IT               2255.97  274515.0      104956.0   
 3  2019     AAPL       IT               1304.76  260174.0       98392.0   
 4  2018     AAPL       IT                748.54  265595.0      101839.0   
 
    Net Income  Earning Per Share    EBITDA  Share Holder Equity  ...  \
 0     99803.0               6.11  130541.0              50672.0  ...   
 1     94680.0               5.61  120233.0              63090.0  ...   
 2     57411.0               3.28   77344.0              65339.0  ...   
 3     55256.0               2.97   76477.0              90488.0  ...   
 4     59531.0               2.98   81801.0             107147.0  ...   
 
    Current Ratio  Debt/Equity Ratio       ROE      ROA      ROI  \
 0         0.8794          

In [5]:
df

Year Company  Category  Market Cap(in B USD)   Revenue  Gross Profit  \
0    2022     AAPL       IT               2066.94  394328.0      170782.0   
1    2021     AAPL       IT               2913.28  365817.0      152836.0   
2    2020     AAPL       IT               2255.97  274515.0      104956.0   
3    2019     AAPL       IT               1304.76  260174.0       98392.0   
4    2018     AAPL       IT                748.54  265595.0      101839.0   
..    ...      ...      ...                   ...       ...           ...   
156  2013     AMZN     LOGI                182.54   74452.0       20271.0   
157  2012     AMZN     LOGI                113.63   61093.0       15122.0   
158  2011     AMZN     LOGI                 78.72   48077.0       10789.0   
159  2010     AMZN     LOGI                 80.79   34204.0        7643.0   
160  2009     AMZN     LOGI                 58.24   24509.0        5531.0   

     Net Income  Earning Per Share    EBITDA  Share Holder Equity  ...  \
0       99803.0             6.1100  130541.0              50672.0  ...   
1       94680.0             5.6100  120233.0              63090.0  ...   
2       57411.0             3.2800   77344.0              65339.0  ...   
3       55256.0             2.9700   76477.0              90488.0  ...   
4       59531.0             2.9800   81801.0             107147.0  ...   
..          ...                ...       ...                  ...  ...   
156       274.0             0.0295    3998.0               9746.0  ...   
157       -39.0            -0.0045    2835.0               8192.0  ...   
158       631.0             0.0685    1945.0               7757.0  ...   
159      1152.0             0.1265    1974.0               6864.0  ...   
160       902.0             0.1020    1507.0               5257.0  ...   

     Current Ratio  Debt/Equity Ratio       ROE      ROA      ROI  \
0           0.8794             2.3695  196.9589  28.2924  66.6994   
1           1.0746             1.9768  150.0713  26.9742  54.9839   
2           1.3636             1.7208   87.8664  17.7256  35.0054   
3           1.5401             1.1940   61.0645  16.3230  30.3113   
4           1.1329             1.0685   55.5601  16.2775  29.6348   
..             ...                ...       ...      ...      ...   
156         1.0716             0.3274    2.8114   0.6823   2.1180   
157         1.1207             0.3765   -0.4761  -0.1198  -0.3459   
158         1.1741             0.0329    8.1346   2.4962   7.8757   
159         1.3254             0.2274   16.7832   6.1286  13.6736   
160         1.3304             0.2007   17.1581   6.5301  17.1581   

     Net Profit Margin  Free Cash Flow per Share  Return on Tangible Equity  \
0              25.3096                    1.3146                   196.9589   
1              25.8818                    1.3261                   150.0713   
2              20.9136                    1.0183                    87.8664   
3              21.2381                   -0.0388                    61.0645   
4              22.4142                    0.7414                    55.5601   
..                 ...                       ...                        ...   
156             0.3680                    0.1748                     3.8641   
157            -0.0638                   -0.1833                    -0.6915   
158             1.3125                   -0.0490                    10.8756   
159             3.3680                   -0.0544                    20.8885   
160             3.6803                    0.3303                    22.4211   

     Number of Employees  Inflation Rate(in US)  
0                 164000                 8.0028  
1                 154000                 4.6979  
2                 147000                 1.2336  
3                 137000                 1.8122  
4                 132000                 2.4426  
..                   ...                    ...  
156               117300                 1.4648  
157          

In [6]:
# Проверим количество пропущенных значений по столбцам
missing_values = df.isnull().sum()

# Посмотрим на пропущенные данные в "Market Cap(in B USD)"
missing_values

Year                                   0
Company                                0
Category                               0
Market Cap(in B USD)                   1
Revenue                                0
Gross Profit                           0
Net Income                             0
Earning Per Share                      0
EBITDA                                 0
Share Holder Equity                    0
Cash Flow from Operating               0
Cash Flow from Investing               0
Cash Flow from Financial Activities    0
Current Ratio                          0
Debt/Equity Ratio                      0
ROE                                    0
ROA                                    0
ROI                                    0
Net Profit Margin                      0
Free Cash Flow per Share               0
Return on Tangible Equity              0
Number of Employees                    0
Inflation Rate(in US)                  0
dtype: int64

In [7]:
# Заполним пропущенное значение медианой столбца "Market Cap(in B USD)"
median_market_cap = df['Market Cap(in B USD)'].median()
df['Market Cap(in B USD)'].fillna(median_market_cap, inplace=True)

# Проверим, что пропущенных значений больше нет
df.isnull().sum()

<ipython-input-7-ed64436666fa>:3: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df['Market Cap(in B USD)'].fillna(median_market_cap, inplace=True)


Year                                   0
Company                                0
Category                               0
Market Cap(in B USD)                   0
Revenue                                0
Gross Profit                           0
Net Income                             0
Earning Per Share                      0
EBITDA                                 0
Share Holder Equity                    0
Cash Flow from Operating               0
Cash Flow from Investing               0
Cash Flow from Financial Activities    0
Current Ratio                          0
Debt/Equity Ratio                      0
ROE                                    0
ROA                                    0
ROI                                    0
Net Profit Margin                      0
Free Cash Flow per Share               0
Return on Tangible Equity              0
Number of Employees                    0
Inflation Rate(in US)                  0
dtype: int64

In [8]:
pip install linearmodels

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.7/1.7 MB 50.9 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 94.5/94.5 kB 7.7 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 43.7/43.7 kB 3.3 MB/s eta 0:00:00


In [9]:
from linearmodels import PanelOLS
import statsmodels.api as sm

In [10]:
df.columns = df.columns.str.strip()


In [11]:
from linearmodels import PanelOLS
import statsmodels.api as sm

# Для начала преобразуем данные в формат панели
# Необходимо, чтобы данные имели MultiIndex с уровнями компании и года
df.set_index(['Company', 'Year'], inplace=True)

# Определяем зависимую и независимые переменные
dependent_var = df['Revenue']
independent_vars = df[['Market Cap(in B USD)', 'Gross Profit', 'Net Income', 'EBITDA',
                         'Share Holder Equity', 'Debt/Equity Ratio', 'ROE', 'Net Profit Margin']]

# Добавляем константу для моделей
independent_vars = sm.add_constant(independent_vars)


In [12]:
#модель с фиксированными эффектами (FE)
fe_model = PanelOLS(dependent_var, independent_vars, entity_effects=True, time_effects=True)
fe_results = fe_model.fit()

fe_results.summary


<class 'linearmodels.compat.statsmodels.Summary'>
"""
                          PanelOLS Estimation Summary                           
================================================================================
Dep. Variable:                Revenue   R-squared:                        0.9743
Estimator:                   PanelOLS   R-squared (Between):              0.8523
No. Observations:                 161   R-squared (Within):               0.9793
Date:                Tue, Oct 29 2024   R-squared (Overall):              0.9087
Time:                        21:14:17   Log-likelihood                   -1680.1
Cov. Estimator:            Unadjusted                                           
                                        F-statistic:                      602.05
Entities:                          12   P-value                           0.0000
Avg Obs:                       13.417   Distribution:                   F(8,127)
Min Obs:                       9.0000                                           
Max Obs:                       15.000   F-statistic (robust):             602.05
                                        P-value                           0.0000
Time periods:                      15   Distribution:                   F(8,127)
Avg Obs:                       10.733                                           
Min Obs:                       2.0000                                           
Max Obs:                       12.000                                           
                                                                                
                                  Parameter Estimates                                   
========================================================================================
                      Parameter  Std. Err.     T-stat    P-value    Lower CI    Upper CI
----------------------------------------------------------------------------------------
const                    1558.2     2209.8     0.7051     0.4820     -2814.6      5930.9
Market Cap(in B USD)    -3.8097     4.5817    -0.8315     0.4073     -12.876      5.2567
Gross Profit             2.3285     0.0741     31.410     0.0000      2.1818      2.4752
Net Income              -0.3619     0.2190    -1.6524     0.1009     -0.7952      0.0715
EBITDA                   0.1046     0.2200     0.4755     0.6353     -0.3308      0.5400
Share Holder Equity     -0.1623     0.0385    -4.2118     0.0000     -0.2386     -0.0860
Debt/Equity Ratio        725.47     651.47     1.1136     0.2676     -563.67      2014.6
ROE                      17.855     27.460     0.6502     0.5167     -36.484      72.193
Net Profit Margin       -39.530     103.53    -0.3818     0.7032     -244.39      165.33
========================================================================================

F-test for Poolability: 32.763
P-value: 0.0000
Distribution: F(25,127)

Included effects: Entity, Time
"""

In [13]:
#модель со случайными эффектами (RE)
from linearmodels.panel import RandomEffects

re_model = RandomEffects(dependent_var, independent_vars)
re_results = re_model.fit()

re_results.summary


<class 'linearmodels.compat.statsmodels.Summary'>
"""
                        RandomEffects Estimation Summary                        
================================================================================
Dep. Variable:                Revenue   R-squared:                        0.9786
Estimator:              RandomEffects   R-squared (Between):              0.8586
No. Observations:                 161   R-squared (Within):               0.9795
Date:                Tue, Oct 29 2024   R-squared (Overall):              0.9121
Time:                        21:14:21   Log-likelihood                   -1695.9
Cov. Estimator:            Unadjusted                                           
                                        F-statistic:                      869.07
Entities:                          12   P-value                           0.0000
Avg Obs:                       13.417   Distribution:                   F(8,152)
Min Obs:                       9.0000                                           
Max Obs:                       15.000   F-statistic (robust):             869.32
                                        P-value                           0.0000
Time periods:                      15   Distribution:                   F(8,152)
Avg Obs:                       10.733                                           
Min Obs:                       2.0000                                           
Max Obs:                       12.000                                           
                                                                                
                                  Parameter Estimates                                   
========================================================================================
                      Parameter  Std. Err.     T-stat    P-value    Lower CI    Upper CI
----------------------------------------------------------------------------------------
const                    3421.0     9479.5     0.3609     0.7187  -1.531e+04   2.215e+04
Market Cap(in B USD)    -2.7955     3.6944    -0.7567     0.4504     -10.094      4.5035
Gross Profit             2.3149     0.0700     33.069     0.0000      2.1766      2.4532
Net Income              -0.3775     0.2083    -1.8128     0.0718     -0.7889      0.0339
EBITDA                   0.0913     0.2111     0.4325     0.6660     -0.3258      0.5084
Share Holder Equity     -0.1553     0.0367    -4.2374     0.0000     -0.2277     -0.0829
Debt/Equity Ratio        144.31     614.56     0.2348     0.8147     -1069.9      1358.5
ROE                      28.222     26.318     1.0724     0.2853     -23.774      80.218
Net Profit Margin       -102.42     99.503    -1.0293     0.3050     -299.01      94.165
========================================================================================
"""

In [ ]:
#объединённую модель (Pooling Model)
pooled_model = PanelOLS(dependent_var, independent_vars)
pooled_results = pooled_model.fit()

pooled_results.summary

<class 'linearmodels.compat.statsmodels.Summary'>
"""
                          PanelOLS Estimation Summary                           
================================================================================
Dep. Variable:                Revenue   R-squared:                        0.9383
Estimator:                   PanelOLS   R-squared (Between):              0.9239
No. Observations:                 161   R-squared (Within):               0.9594
Date:                Wed, Sep 25 2024   R-squared (Overall):              0.9383
Time:                        23:32:34   Log-likelihood                   -1841.7
Cov. Estimator:            Unadjusted                                           
                                        F-statistic:                      288.97
Entities:                          12   P-value                           0.0000
Avg Obs:                       13.417   Distribution:                   F(8,152)
Min Obs:                       9.0000                                           
Max Obs:                       15.000   F-statistic (robust):             288.97
                                        P-value                           0.0000
Time periods:                      15   Distribution:                   F(8,152)
Avg Obs:                       10.733                                           
Min Obs:                       2.0000                                           
Max Obs:                       12.000                                           
                                                                                
                                  Parameter Estimates                                   
========================================================================================
                      Parameter  Std. Err.     T-stat    P-value    Lower CI    Upper CI
----------------------------------------------------------------------------------------
const                 1.257e+04     3372.2     3.7286     0.0003      5911.1   1.924e+04
Market Cap(in B USD)    -13.294     8.1331    -1.6345     0.1042     -29.362      2.7748
Gross Profit             2.1101     0.1200     17.584     0.0000      1.8730      2.3471
Net Income              -0.8811     0.4239    -2.0788     0.0393     -1.7185     -0.0437
EBITDA                   1.2154     0.3595     3.3812     0.0009      0.5052      1.9256
Share Holder Equity     -0.2196     0.0556    -3.9519     0.0001     -0.3294     -0.1098
Debt/Equity Ratio       -2940.6     1080.0    -2.7228     0.0072     -5074.4     -806.87
ROE                      152.78     55.370     2.7593     0.0065      43.388      262.18
Net Profit Margin       -896.63     179.83    -4.9861     0.0000     -1251.9     -541.35
========================================================================================


"""

1. Общие характеристики модели:
Dep. Variable (Зависимая переменная): Revenue — это переменная, которую мы пытаемся предсказать.
R-squared: 0.9383 — это коэффициент детерминации, который показывает долю объясненной дисперсии зависимой переменной. Значение 0.9383 означает, что модель объясняет 93.83% вариации доходов компании.
R-squared (Between): 0.9239 — это коэффициент детерминации между объектами (компаниями), показывающий, насколько хорошо модель объясняет различия между объектами.
R-squared (Within): 0.9594 — это коэффициент, показывающий, насколько хорошо модель объясняет различия в динамике во времени для одного объекта.
R-squared (Overall): 0.9383 — общий показатель, показывающий, как модель объясняет общую вариацию данных.
No. Observations (Число наблюдений): 161 — общее количество наблюдений в данных.
Entities (Количество объектов): 12 — количество компаний.
Time periods: 15 — количество временных периодов (лет).
F-statistic: 288.97 — значение F-критерия, который показывает, насколько значимы все регрессоры (независимые переменные) совместно. P-value равен 0.0000, что свидетельствует о высокой значимости модели.
2. Параметры модели (Parameter Estimates):
Это ключевая часть отчета, которая показывает, как независимые переменные влияют на зависимую переменную.

Столбцы:
Parameter (Параметр): Оценки коэффициентов для каждой независимой переменной и константы.
Std. Err. (Стандартная ошибка): Показывает неопределенность в оценке коэффициента. Чем меньше стандартная ошибка, тем точнее оценка.
T-stat: T-статистика для проверки гипотезы, что коэффициент равен нулю. Чем больше t-статистика по модулю, тем более значима переменная.
P-value: P-значение для t-статистики. Если P-value меньше 0.05, то коэффициент значим на уровне 5%.
Lower CI, Upper CI (Доверительный интервал): Нижняя и верхняя границы 95% доверительного интервала для коэффициента.
Разбор параметров:
const (Константа): 12570.0 — базовое значение дохода компании, когда все независимые переменные равны нулю.
T-stat = 3.7286, P-value = 0.0003: Значима на уровне 5%.
Market Cap(in B USD): -13.294 — при увеличении рыночной капитализации на 1 миллиард долларов, доход компании уменьшается на 13.294 единицы. Однако переменная незначима на уровне 5% (P-value = 0.1042).
Gross Profit: 2.1101 — валовая прибыль сильно и положительно влияет на доход. При увеличении валовой прибыли на единицу, доход увеличивается на 2.11. Переменная значима (P-value = 0.0000).
Net Income: -0.8811 — чистая прибыль оказывает отрицательное влияние на доход (на каждый доллар чистой прибыли доход уменьшается на 0.88), и она значима (P-value = 0.0393).
EBITDA: 1.2154 — увеличение EBITDA на единицу увеличивает доход на 1.2154. Переменная значима (P-value = 0.0009).
Share Holder Equity: -0.2196 — увеличение доли акционеров на единицу уменьшает доход на 0.2196. Переменная значима (P-value = 0.0001).
Debt/Equity Ratio: -2940.6 — соотношение долга к капиталу сильно отрицательно влияет на доход (уменьшает на 2940.6 единиц при увеличении отношения на единицу). Переменная значима (P-value = 0.0072).
ROE: 152.78 — рост рентабельности капитала на 1% увеличивает доход на 152.78. Переменная значима (P-value = 0.0065).
Net Profit Margin: -896.63 — увеличение чистой маржи снижает доход на 896.63. Переменная значима (P-value = 0.0000).
3. Выводы:
Модель объясняет значительную часть вариации доходов компании (93.83%).
Некоторые переменные, такие как валовая прибыль, EBITDA, ROE, и чистая маржа, оказывают сильное влияние на доход.
Рыночная капитализация незначима в этой модели (P-value > 0.05), но другие переменные, такие как валовая прибыль и чистая маржа, играют важную роль.

In [14]:
# Перестроим модели, исключив переменную "Market Cap(in B USD)"
independent_vars_reduced = independent_vars.drop(columns=['Market Cap(in B USD)'])

# Модель с фиксированными эффектами (FE)
fe_model_reduced = PanelOLS(dependent_var, independent_vars_reduced, entity_effects=True)
fe_results_reduced = fe_model_reduced.fit()

# Модель со случайными эффектами (RE)
re_model_reduced = RandomEffects(dependent_var, independent_vars_reduced)
re_results_reduced = re_model_reduced.fit()

# Пула (Pooling Model)
pooled_model_reduced = PanelOLS(dependent_var, independent_vars_reduced)
pooled_results_reduced = pooled_model_reduced.fit()

# Вывод результатов
print(fe_results_reduced.summary)
print(re_results_reduced.summary)
print(pooled_results_reduced.summary)


                          PanelOLS Estimation Summary                           
Dep. Variable:                Revenue   R-squared:                        0.9794
Estimator:                   PanelOLS   R-squared (Between):              0.8550
No. Observations:                 161   R-squared (Within):               0.9794
Date:                Tue, Oct 29 2024   R-squared (Overall):              0.9102
Time:                        21:14:27   Log-likelihood                   -1692.3
Cov. Estimator:            Unadjusted                                           
                                        F-statistic:                      963.83
Entities:                          12   P-value                           0.0000
Avg Obs:                       13.417   Distribution:                   F(7,142)
Min Obs:                       9.0000                                           
Max Obs:                       15.000   F-statistic (robust):             963.83
                            